File: SWEA/D3/3307. 최장 증가 부분 수열/최장 증가 부분 수열.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int LIS(const vector<int>& arr){
    int n = arr.size();
    vector<int> dp(n, 1);

    for(int i = 0; i < n; i++){
        for(int j = 0; j < i; j++){
            if(arr[j] < arr[i]){
                dp[i] = max(dp[i], dp[j] + 1);
            }
        }
    }

    return *max_element(dp.begin(), dp.end());
}

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    int T;
    cin >> T;

    for(int t = 1; t <= T; t++){
        int N;
        cin >> N;
        vector<int> arr(N, 0);

        for(int n = 0; n < N; n++){
            cin >> arr[n];
        }

        cout << '#' << t << ' ' << LIS(arr) << '\n';
    }
    return 0;
}

File: SWEA/D3/3307. 최장 증가 부분 수열/최장 증가 부분 수열.py
def solve():
    N = int(input())
    arr = list(map(int, input().split()))

    def lis(N, arr):
        if N == 0:
            return 0

        # dp[i] : arr[i]를 마지막 원소로 가지는 LIS의 길이
        dp = [1] * N

        for i in range(N):
            for j in range(i):
                if arr[j] < arr[i]:
                    dp[i] = max(dp[i], dp[j] + 1)

        return max(dp)

    return lis(N, arr)

T = int(input())
for tc in range(1, T+1):
    print(f'#{tc} {solve()}')